java-checks-test-sources/src/main/java/symbolicexecution/checks/S6377_XmlValidatedSighatureCheck.java

```diff
@@ -0,0 +1,85 @@
+package symbolicexecution.checks;
+
+import javax.xml.crypto.KeySelector;
+import javax.xml.crypto.dsig.XMLSignature;
+import javax.xml.crypto.dsig.XMLSignatureFactory;
+import javax.xml.crypto.dsig.dom.DOMValidateContext;
+import org.w3c.dom.Node;
+
+abstract class S3677_XmlValidatedSighatureCheck {
+  private final XMLSignatureFactory fac = XMLSignatureFactory.getInstance("DOM");
+  private DOMValidateContext valContextField;
+  private Object field;
+
+  void compliant(Node node, KeySelector selector) throws Exception {
+    DOMValidateContext valContext = new DOMValidateContext(selector, node);
+    valContext.setProperty("org.jcp.xml.dsig.secureValidation", Boolean.TRUE); // Compliant
+
+    XMLSignature signature = fac.unmarshalXMLSignature(valContext);
+
+    signature.validate(valContext);
+
+    // -------------------
+
+    valContextField = new DOMValidateContext(selector, node);
+    valContextField.setProperty("org.jcp.xml.dsig.secureValidation", Boolean.FALSE); // Compliant - field...
+
+    XMLSignature signature2 = fac.unmarshalXMLSignature(valContextField);
+    signature2.validate(valContextField);
+  }
+
+  void compliant_2(Node node, KeySelector selector) throws Exception {
+    DOMValidateContext valContext = new DOMValidateContext(selector, node);
+    valContext.setProperty("org.jcp.xml.dsig.secureValidation", foo()); // Compliant - no idea of what it could be
+
+    XMLSignature signature = fac.unmarshalXMLSignature(valContext);
+
+    signature.validate(valContext);
+  }
+
+  private void coverage(DOMValidateContext valContext) {
+    field = new Object();
+
+    Object[] arr = new Object[1];
+    arr[0] = field;
+
+    valContext.setProperty("some.other.feature", field);
+  }
+
+  abstract boolean foo();
+
+  void noncompliant_1(Node node, KeySelector selector) throws Exception {
+    DOMValidateContext valContext = new DOMValidateContext(selector, node);  // Noncompliant {{Use the secure validation mode when validating this XML signature.}}
+    valContext.setProperty("org.jcp.xml.dsig.secureValidation", Boolean.FALSE);
+
+    XMLSignature signature = fac.unmarshalXMLSignature(valContext);
+
+    signature.validate(valContext);
+  }
+
+  void noncompliant_2(Node node, KeySelector selector) throws Exception {
+    DOMValidateContext valContext = new DOMValidateContext(selector, node); // Noncompliant {{Use the secure validation mode when validating this XML signature.}}
+
+    XMLSignature signature = fac.unmarshalXMLSignature(valContext);
+
+    signature.validate(valContext);
+  }
+
+  void noncompliant_3(Node node, KeySelector selector) throws Exception {
+    DOMValidateContext valContext = new DOMValidateContext(selector, node);
+    valContext.setProperty("org.jcp.xml.dsig.secureValidation", Boolean.TRUE);
+    valContext.setProperty("org.jcp.xml.dsig.secureValidation", Boolean.FALSE); // Noncompliant {{Use the secure validation mode when validating this XML signature.}}
```

Comment:
We will report on the `setProperty` only if the constraint changes after the constructor, and not when the constraint was already "DISABLED":

```
DOMValidateContext valContext = new DOMValidateContext(selector, node);  // Noncompliant {{Use the secure validation mode when validating this XML signature.}}
valContext.setProperty("org.jcp.xml.dsig.secureValidation", Boolean.FALSE); // Issue could be here?
```

We should either always report on `setProperty` (when present), or always on the constructor (with secondary location on the problematic `setProperty`). We use the second solution for S2755.

Reply:
I think it's better to always highlight the setter when it is explicit to FALSE, and the constructor if there is no such call. I don't really like the secondary approach since the real place when we should fix things is in the setter.